# Introduction
In this project, we will explore a dataset of used cars from eBay Kleinanzeigen, the classifieds section of the German eBay website. The dataset contains various attributes of the cars listed, such as their make, model, price, odometer reading, registration year, and more. Our objective is to clean the data and perform some initial analysis on it. The data dictionary provided with data is as follows:

* dateCrawled - When this ad was first crawled. All field-values are taken from this date.
* name - Name of the car.
* seller - Whether the seller is private or a dealer.
* offerType - The type of listing
* price - The price on the ad to sell the car.
* abtest - Whether the listing is included in an A/B test.
* vehicleType - The vehicle Type.
* yearOfRegistration - The year in which which year the car was first registered.
* gearbox - The transmission type.
* powerPS - The power of the car in PS.
* model - The car model name.
* kilometer - How many kilometers the car has driven.
* monthOfRegistration - The month in which which year the car was first registered.
* fuelType - What type of fuel the car uses.
* brand - The brand of the car.
* notRepairedDamage - If the car has a damage which is not yet repaired.
* dateCreated - The date on which the eBay listing was created.
* nrOfPictures - The number of pictures in the ad.
* postalCode - The postal code for the location of the vehicle.
* lastSeenOnline - When the crawler saw this ad last online.

In [1]:
# Importing necessary libraries
import pandas as pd
import numpy as np

In [2]:
# Reading the dataset
autos = pd.read_csv('autos.csv')

In [3]:
autos

dateCrawled                                               name  \
0      2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1      2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2      2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3      2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4      2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_T�V_neu_ist_sehr_gepfleg...   
...                    ...                                                ...   
49995  2016-03-27 14:38:19   Audi_Q5_3.0_TDI_qu._S_tr.__Navi__Panorama__Xenon   
49996  2016-03-28 10:50:25  Opel_Astra_F_Cabrio_Bertone_Edition___T�V_neu+...   
49997  2016-04-02 14:44:48                     Fiat_500_C_1.2_Dualogic_Lounge   
49998  2016-03-08 19:25:42                 Audi_A3_2.0_TDI_Sportback_Ambition   
49999  2016-03-14 00:42:12                                Opel_Vectra_1.6_16V   

       seller offerType    price   abtest vehicleType  yearOfRegistration  \
0      privat   Angebot   $5,000  control         bus                2004   
1      privat   Angebot   $8,500  control   limousine                1997   
2      privat   Angebot   $8,990     test   limousine                2009   
3      privat   Angebot   $4,350  control  kleinwagen                2007   
4      privat   Angebot   $1,350     test       kombi                2003   
...       ...       ...      ...      ...         ...                 ...   
49995  privat   Angebot  $24,900  control   limousine                2011   
49996  privat   Angebot   $1,980  control      cabrio                1996   
49997  privat   Angebot  $13,200     test      cabrio                2014   
49998  privat   Angebot  $22,900  control       kombi                2013   
49999  privat   Angebot   $1,250  control   limousine                1996   

         gearbox  powerPS   model   odometer  monthOfRegistration fuelType  \
0        manuell      158  andere  150,000km                    3      lpg   
1      automatik      286     7er  150,000km                    6   benzin   
2        manuell      102    golf   70,000km                    7   benzin   
3      automatik       71  fortwo   70,000km                    6   benzin   
4        manuell        0   focus  150,000km                    7   benzin   
...          ...      ...     ...        ...                  ...      ...   
49995  automatik      239      q5  100,000km                    1   diesel   
49996    manuell       75   astra  150,000km                    5   benzin   
49997  automatik       69     500    5,000km                   11   benzin   
49998    manuell      150      a3   40,000km                   11   diesel   
49999    manuell      101  vectra  150,000km                    1   benzin   

            brand notRepairedDamage          dateCreated  nrOfPictures  \
0         peugeot              nein  2016-03-26 00:00:00             0   
1             bmw              nein  2016-04-04 00:00:00             0   
2      volkswagen              nein  2016-03-26 00:00:00             0   
3           smart              nein  2016-03-12 00:00:00             0   
4            ford              nein  2016-04-01 00:00:00             0   
...           ...               ...                  ...           ...   
49995        audi              nein  2016-03-27 00:00:00             0   
49996        opel              nein  2016-03-28 00:00:00             0   
49997        fiat              nein  2016-04-02 00:00:00             0   
49998        audi              nein  2016-03-08 00:00:00             0   
49999        opel              nein  2016-03-13 00:00:00             0   

       postalCode             lastSeen  
0           79588  2016-04-06 06:45:54  
1           71034  2016-04-06 14:45:08  
2           35394  2016-04-06 20:15:37  
3           33729  2016-03-15 03:16:28  
4           39218  2016-04-01 14:38:50  
...           ...                  ...  
4999

In [4]:
# Displaying information about the dataset
autos.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 20 columns):
 #   Column               Non-Null Count  Dtype 
---  ------               --------------  ----- 
 0   dateCrawled          50000 non-null  object
 1   name                 50000 non-null  object
 2   seller               50000 non-null  object
 3   offerType            50000 non-null  object
 4   price                50000 non-null  object
 5   abtest               50000 non-null  object
 6   vehicleType          44905 non-null  object
 7   yearOfRegistration   50000 non-null  int64 
 8   gearbox              47320 non-null  object
 9   powerPS              50000 non-null  int64 
 10  model                47242 non-null  object
 11  odometer             50000 non-null  object
 12  monthOfRegistration  50000 non-null  int64 
 13  fuelType             45518 non-null  object
 14  brand                50000 non-null  object
 15  notRepairedDamage    40171 non-null  object
 16  date

In [5]:
# Displaying the first few rows of the dataset
autos.head()

dateCrawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_T�V_neu_ist_sehr_gepfleg...   

   seller offerType   price   abtest vehicleType  yearOfRegistration  \
0  privat   Angebot  $5,000  control         bus                2004   
1  privat   Angebot  $8,500  control   limousine                1997   
2  privat   Angebot  $8,990     test   limousine                2009   
3  privat   Angebot  $4,350  control  kleinwagen                2007   
4  privat   Angebot  $1,350     test       kombi                2003   

     gearbox  powerPS   model   odometer  monthOfRegistration fuelType  \
0    manuell      158  andere  150,000km                    3      lpg   
1  automatik      286     7er  150,000km                    6   benzin   
2    manuell      102    golf   70,000km                    7   benzin   
3  automatik       71  fortwo   70,000km                    6   benzin   
4    manuell        0   focus  150,000km                    7   benzin   

        brand notRepairedDamage          dateCreated  nrOfPictures  \
0     peugeot              nein  2016-03-26 00:00:00             0   
1         bmw              nein  2016-04-04 00:00:00             0   
2  volkswagen              nein  2016-03-26 00:00:00             0   
3       smart              nein  2016-03-12 00:00:00             0   
4        ford              nein  2016-04-01 00:00:00             0   

   postalCode             lastSeen  
0       79588  2016-04-06 06:45:54  
1       71034  2016-04-06 14:45:08  
2       35394  2016-04-06 20:15:37  
3       33729  2016-03-15 03:16:28  
4       39218  2016-04-01 14:38:50

# Observations
* The dataset comprises 50,000 entries and 20 columns.
* There are several columns with missing values, such as 'vehicleType', 'gearbox', 'model', 'fuelType', and 'notRepairedDamage'.
* Most columns are of object type (strings), while some are of integer type.
* The 'price' and 'odometer' columns are stored as objects, which might require conversion to numeric types for analysis.

# Cleaning Column Names

In [6]:
# Printing an array containing the column names
print(autos.columns)

Index(['dateCrawled', 'name', 'seller', 'offerType', 'price', 'abtest',
       'vehicleType', 'yearOfRegistration', 'gearbox', 'powerPS', 'model',
       'odometer', 'monthOfRegistration', 'fuelType', 'brand',
       'notRepairedDamage', 'dateCreated', 'nrOfPictures', 'postalCode',
       'lastSeen'],
      dtype='object')


In [7]:
modified_columns=['date_crawled', 'name', 'seller', 'offer_type', 'price', 'abtest', 'vehicle_type', 
                  'registration_year', 'gearbox', 'power_ps', 'model', 'odometer', 'registration_month', 
                  'fuel_type', 'brand', 'unrepaired_damage', 'ad_created', 'num_photos', 'postal_code',
                  'last_seen']


In [8]:
# Modifying the column names of the DataFrame
autos.columns=modified_columns

In [9]:
# Displaying the first few rows to verify the changes
autos.head()

date_crawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_T�V_neu_ist_sehr_gepfleg...   

   seller offer_type   price   abtest vehicle_type  registration_year  \
0  privat    Angebot  $5,000  control          bus               2004   
1  privat    Angebot  $8,500  control    limousine               1997   
2  privat    Angebot  $8,990     test    limousine               2009   
3  privat    Angebot  $4,350  control   kleinwagen               2007   
4  privat    Angebot  $1,350     test        kombi               2003   

     gearbox  power_ps   model   odometer  registration_month fuel_type  \
0    manuell       158  andere  150,000km                   3       lpg   
1  automatik       286     7er  150,000km                   6    benzin   
2    manuell       102    golf   70,000km                   7    benzin   
3  automatik        71  fortwo   70,000km                   6    benzin   
4    manuell         0   focus  150,000km                   7    benzin   

        brand unrepaired_damage           ad_created  num_photos  postal_code  \
0     peugeot              nein  2016-03-26 00:00:00           0        79588   
1         bmw              nein  2016-04-04 00:00:00           0        71034   
2  volkswagen              nein  2016-03-26 00:00:00           0        35394   
3       smart              nein  2016-03-12 00:00:00           0        33729   
4        ford              nein  2016-04-01 00:00:00           0        39218   

             last_seen  
0  2016-04-06 06:45:54  
1  2016-04-06 14:45:08  
2  2016-04-06 20:15:37  
3  2016-03-15 03:16:28  
4  2016-04-01 14:38:50

In this step, we modified the column names of the autos DataFrame for improved readability and adherence to Python conventions. Here are the changes we made:

* yearOfRegistration to registration_year: Renamed to make it clearer that this column represents the registration year of the vehicle.
* monthOfRegistration to registration_month: Renamed to indicate that this column represents the registration month of the vehicle.
* notRepairedDamage to unrepaired_damage: Renamed for consistency and to follow snake_case convention.
* dateCreated to ad_created: Renamed to better reflect that this column represents the date when the ad was created.
* Changed the remaining column names from camelcase to snakecase: This was done to maintain consistency throughout the column names and to adhere to Python naming conventions.

By making these changes, the column names become more descriptive and easier to understand, which enhances the readability and usability of the DataFrame.







#  Initial Exploration and Cleaning

In [10]:
# Descriptive statistics for all columns
autos.describe(include='all')

date_crawled         name  seller offer_type  price abtest  \
count                 50000        50000   50000      50000  50000  50000   
unique                48213        38754       2          2   2357      2   
top     2016-03-09 11:54:38  Ford_Fiesta  privat    Angebot     $0   test   
freq                      3           78   49999      49999   1421  25756   
mean                    NaN          NaN     NaN        NaN    NaN    NaN   
std                     NaN          NaN     NaN        NaN    NaN    NaN   
min                     NaN          NaN     NaN        NaN    NaN    NaN   
25%                     NaN          NaN     NaN        NaN    NaN    NaN   
50%                     NaN          NaN     NaN        NaN    NaN    NaN   
75%                     NaN          NaN     NaN        NaN    NaN    NaN   
max                     NaN          NaN     NaN        NaN    NaN    NaN   

       vehicle_type  registration_year  gearbox      power_ps  model  \
count         44905       50000.000000    47320  50000.000000  47242   
unique            8                NaN        2           NaN    245   
top       limousine                NaN  manuell           NaN   golf   
freq          12859                NaN    36993           NaN   4024   
mean            NaN        2005.073280      NaN    116.355920    NaN   
std             NaN         105.712813      NaN    209.216627    NaN   
min             NaN        1000.000000      NaN      0.000000    NaN   
25%             NaN        1999.000000      NaN     70.000000    NaN   
50%             NaN        2003.000000      NaN    105.000000    NaN   
75%             NaN        2008.000000      NaN    150.000000    NaN   
max             NaN        9999.000000      NaN  17700.000000    NaN   

         odometer  registration_month fuel_type       brand unrepaired_damage  \
count       50000        50000.000000     45518       50000             40171   
unique         13                 NaN         7          40                 2   
top     150,000km                 NaN    benzin  volkswagen              nein   
freq        32424                 NaN     30107       10687             35232   
mean          NaN            5.723360       NaN         NaN               NaN   
std           NaN            3.711984       NaN         NaN               NaN   
min           NaN            0.000000       NaN         NaN               NaN   
25%           NaN            3.000000       NaN         NaN               NaN   
50%           NaN            6.000000       NaN         NaN               NaN   
75%           NaN            9.000000       NaN         NaN               NaN   
max           NaN           12.000000       NaN         NaN               NaN   

                 ad_created  num_photos   postal_code            last_seen  
count                 50000     50000.0  50000.000000                50000  
unique                   76         NaN           NaN                39481  
top     2016-04-03 00:00:00         NaN           NaN  2016-04-07 06:17:27  
freq                   1946         NaN           NaN                    8  
mean                    NaN         0.0  50813.627300                  NaN  
std                     NaN         0.0  25779.747957                  NaN  
min                     NaN         0.0   1067.000000                  NaN  
25%                     NaN         0.0  30451.000000                  NaN  
50%                     NaN         0.0  49577.000000                  NaN  
75%                     NaN         0.0  71540.000000                  NaN  
max                     NaN         0.0  99998.000000                  NaN

Columns with Mostly One Value:
* seller: This column has mostly one value ("privat") and only a few instances of another value ("gewerblich"). It might be a candidate for dropping since it doesn't provide much variability.
* offer_type: Similarly, this column has mostly one value ("Angebot") and only a few instances of another value ("Gesuch").

Columns Needing More Investigation:
* price: The 'price' column contains numeric data but is currently stored as text due to the presence of '$' and potentially other non-numeric characters. It needs to be cleaned and converted to numeric format for analysis.
* odometer: Similar to the 'price' column, the 'odometer' column contains numeric data stored as text with the unit 'km'. It needs to be cleaned and converted to numeric format.
* registration_year: The 'registration_year' column has a minimum value of 1000 and a maximum value of 9999, which seem unrealistic. Further investigation is needed to identify and handle outliers and erroneous data.
* num_photos: The 'num_photos' column has a value of 0 for all entries, which suggests that it may not contain useful information and can be dropped.

Numeric Data Stored as Text:
* price: Numeric data stored as text due to the presence of '$' symbol.
* odometer: Numeric data stored as text with 'km' suffix.

In [11]:
# Dropping the rows that won't be required for any anlaysis 
autos=autos.drop(['seller','offer_type','num_photos'],axis=1)

To further investigate the columns that require more investigation, let's take a closer look at the 'registration_year' column to identify any outliers or erroneous data:



In [12]:
# Checking unique values and their counts in the 'registration_year' column
autos['registration_year'].value_counts().sort_index()

1000    1
1001    1
1111    1
1500    1
1800    2
       ..
6200    1
8888    1
9000    2
9996    1
9999    4
Name: registration_year, Length: 97, dtype: int64

It seems like there are several unrealistic values in the 'registration_year' column, such as 1000, 1001, 1111, 1500, 1800, 6200, 8888, 9000, 9996, and 9999. These values are likely errors or outliers and need to be addressed.


In [13]:
# Cleaning the 'price' column
autos['price']=autos['price'].str.replace('$','',regex=False).str.replace(',','',regex=False).astype(float)

In [14]:
# Cleaning the 'odometer' column and renaming it to 'odometer_km'
autos['odometer']=autos['odometer'].str.replace('km','',regex=False).str.replace(',','',regex=False).astype(float)
autos.rename(columns={'odometer':'odometer_km'},inplace=True)

In [15]:
# Displaying the first few rows to verify the changes
autos.head()

date_crawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_T�V_neu_ist_sehr_gepfleg...   

    price   abtest vehicle_type  registration_year    gearbox  power_ps  \
0  5000.0  control          bus               2004    manuell       158   
1  8500.0  control    limousine               1997  automatik       286   
2  8990.0     test    limousine               2009    manuell       102   
3  4350.0  control   kleinwagen               2007  automatik        71   
4  1350.0     test        kombi               2003    manuell         0   

    model  odometer_km  registration_month fuel_type       brand  \
0  andere     150000.0                   3       lpg     peugeot   
1     7er     150000.0                   6    benzin         bmw   
2    golf      70000.0                   7    benzin  volkswagen   
3  fortwo      70000.0                   6    benzin       smart   
4   focus     150000.0                   7    benzin        ford   

  unrepaired_damage           ad_created  postal_code            last_seen  
0              nein  2016-03-26 00:00:00        79588  2016-04-06 06:45:54  
1              nein  2016-04-04 00:00:00        71034  2016-04-06 14:45:08  
2              nein  2016-03-26 00:00:00        35394  2016-04-06 20:15:37  
3              nein  2016-03-12 00:00:00        33729  2016-03-15 03:16:28  
4              nein  2016-04-01 00:00:00        39218  2016-04-01 14:38:50

# Exploring the Odometer and Price Columns

Let's explore the 'odometer_km' and 'price' columns, identify outliers, and remove them if necessary:

In [16]:
# Exploring the 'odometer_km' column
print("Number of unique values in odometer_km column:", autos['odometer_km'].unique().shape[0])
print("\nDescriptive statistics for odometer_km column:\n", autos['odometer_km'].describe())
print("\nTop 5 most common odometer_km values:\n", autos['odometer_km'].value_counts().head())

Number of unique values in odometer_km column: 13

Descriptive statistics for odometer_km column:
 count     50000.000000
mean     125732.700000
std       40042.211706
min        5000.000000
25%      125000.000000
50%      150000.000000
75%      150000.000000
max      150000.000000
Name: odometer_km, dtype: float64

Top 5 most common odometer_km values:
 150000.0    32424
125000.0     5170
100000.0     2169
90000.0      1757
80000.0      1436
Name: odometer_km, dtype: int64


No outliers were identified in the 'odometer_km' column based on the provided information. The values appear reasonable and consistent, with no extremely high or low values that would warrant removal.



In [17]:
# Exploring the 'price' column
print("Number of unique values in price column:", autos['price'].unique().shape[0])
print("\nDescriptive statistics for price column:\n", autos['price'].describe())
print("\nTop 15 most common price values:\n", autos['price'].value_counts().sort_index(ascending=False).head(15))

Number of unique values in price column: 2357

Descriptive statistics for price column:
 count    5.000000e+04
mean     9.840044e+03
std      4.811044e+05
min      0.000000e+00
25%      1.100000e+03
50%      2.950000e+03
75%      7.200000e+03
max      1.000000e+08
Name: price, dtype: float64

Top 15 most common price values:
 99999999.0    1
27322222.0    1
12345678.0    3
11111111.0    2
10000000.0    1
3890000.0     1
1300000.0     1
1234566.0     1
999999.0      2
999990.0      1
350000.0      1
345000.0      1
299000.0      1
295000.0      1
265000.0      1
Name: price, dtype: int64


In our samples, the prices of all the deal were provided. There are only 2357 unique values, this seems to be the result of people's tendency to round prices on the site. The lowest value is $0.00 (with 1421 deals) which is about 2.8% of total deals so we can remove them; and the highest value is 99999999. Let's have a closer look in this row:

In [18]:
autos[autos['price'] == 99999999]

date_crawled                         name       price   abtest  \
39705  2016-03-22 14:58:27  Tausch_gegen_gleichwertiges  99999999.0  control   

      vehicle_type  registration_year    gearbox  power_ps     model  \
39705    limousine               1999  automatik       224  s_klasse   

       odometer_km  registration_month fuel_type          brand  \
39705     150000.0                   9    benzin  mercedes_benz   

      unrepaired_damage           ad_created  postal_code            last_seen  
39705               NaN  2016-03-22 00:00:00        73525  2016-04-06 05:15:30

The price seems to be strange because, this is not a luxury type of car, and the registration year is 1999, which is not too old to be antique. So we will drop this row.

From our results, there is a hug jump of price from 350000 usd to 999990 usd. And the number of transactions between 1 usd - 350000 usd is:

In [19]:
per_1_350000 = (autos[autos['price'].between(1,350000)].shape[0]/50000)*100
print(round(per_1_350000,2),'%')

97.13 %


Therefore, we will drop these rows as follows:

In [20]:
# drop rows with price = 0 usd:
autos.drop(autos.index[autos['price']==0].tolist(),axis = 0, inplace = True)
# drop rows with price > 350000 usd:
list_index_350000 = autos.index[autos['price'] > 350000].tolist() 
autos.drop(list_index_350000, axis = 0, inplace = True) 
autos['price'].describe()

count     48565.000000
mean       5888.935591
std        9059.854754
min           1.000000
25%        1200.000000
50%        3000.000000
75%        7490.000000
max      350000.000000
Name: price, dtype: float64

# Exploring the Date Columns


To explore the date columns in the dataset, we'll focus on the following columns:

* date_crawled: Added by the crawler.
* last_seen: Added by the crawler.
* ad_created: From the website.
* registration_month: From the website.
* registration_year: From the website.

We'll start by examining the format of the values in the date_crawled, ad_created, and last_seen columns. These columns represent full timestamp values, which we'll extract to understand the date range covered by the data.

In [21]:
# Extracting the first 10 character date values from the 'date_crawled' column
print(autos['date_crawled'].str[:10])

0        2016-03-26
1        2016-04-04
2        2016-03-26
3        2016-03-12
4        2016-04-01
            ...    
49995    2016-03-27
49996    2016-03-28
49997    2016-04-02
49998    2016-03-08
49999    2016-03-14
Name: date_crawled, Length: 48565, dtype: object


In [22]:
# Distribution of values in the 'date_crawled' column
autos['date_crawled'].str[:10].value_counts(normalize=True,dropna=False).sort_index()

2016-03-05    0.025327
2016-03-06    0.014043
2016-03-07    0.036014
2016-03-08    0.033296
2016-03-09    0.033090
2016-03-10    0.032184
2016-03-11    0.032575
2016-03-12    0.036920
2016-03-13    0.015670
2016-03-14    0.036549
2016-03-15    0.034284
2016-03-16    0.029610
2016-03-17    0.031628
2016-03-18    0.012911
2016-03-19    0.034778
2016-03-20    0.037887
2016-03-21    0.037373
2016-03-22    0.032987
2016-03-23    0.032225
2016-03-24    0.029342
2016-03-25    0.031607
2016-03-26    0.032204
2016-03-27    0.031092
2016-03-28    0.034860
2016-03-29    0.034099
2016-03-30    0.033687
2016-03-31    0.031834
2016-04-01    0.033687
2016-04-02    0.035478
2016-04-03    0.038608
2016-04-04    0.036487
2016-04-05    0.013096
2016-04-06    0.003171
2016-04-07    0.001400
Name: date_crawled, dtype: float64

Observations:
* The data was crawled over a period spanning from March 5, 2016, to April 7, 2016.
* The distribution of dates appears relatively uniform, with no significant spikes or drops, indicating consistent data collection throughout the period.
* The dates near the beginning and end of the range have slightly lower percentages, suggesting partial data collection on those days.

In [23]:
# Extracting the first 10 character date values from the 'ad_created' column
print(autos['ad_created'].str[:10])

0        2016-03-26
1        2016-04-04
2        2016-03-26
3        2016-03-12
4        2016-04-01
            ...    
49995    2016-03-27
49996    2016-03-28
49997    2016-04-02
49998    2016-03-08
49999    2016-03-13
Name: ad_created, Length: 48565, dtype: object


In [24]:
# Distribution of values in the 'ad_created' column
autos['ad_created'].str[:10].value_counts(normalize=True,dropna=False).sort_index()

2015-06-11    0.000021
2015-08-10    0.000021
2015-09-09    0.000021
2015-11-10    0.000021
2015-12-05    0.000021
                ...   
2016-04-03    0.038855
2016-04-04    0.036858
2016-04-05    0.011819
2016-04-06    0.003253
2016-04-07    0.001256
Name: ad_created, Length: 76, dtype: float64

Observations:

* The ads were created over a wide range of dates, with the earliest ad creation date recorded as August 10, 2015, and the latest as April 7, 2016.
* There is significant variability in the distribution of ad creation dates, with some dates having only a few ads created, while others have a higher proportion of ads.
* The distribution is right-skewed, with a larger number of ads created in more recent months, particularly in April 2016.






In [25]:
# Extracting the first 10 character date values from the 'last_seen' column
print(autos['last_seen'].str[:10])

0        2016-04-06
1        2016-04-06
2        2016-04-06
3        2016-03-15
4        2016-04-01
            ...    
49995    2016-04-01
49996    2016-04-02
49997    2016-04-04
49998    2016-04-05
49999    2016-04-06
Name: last_seen, Length: 48565, dtype: object


In [26]:
# Distribution of values in the 'last_seen' column
autos['last_seen'].str[:10].value_counts(normalize=True,dropna=False).sort_index()

2016-03-05    0.001071
2016-03-06    0.004324
2016-03-07    0.005395
2016-03-08    0.007413
2016-03-09    0.009595
2016-03-10    0.010666
2016-03-11    0.012375
2016-03-12    0.023783
2016-03-13    0.008895
2016-03-14    0.012602
2016-03-15    0.015876
2016-03-16    0.016452
2016-03-17    0.028086
2016-03-18    0.007351
2016-03-19    0.015834
2016-03-20    0.020653
2016-03-21    0.020632
2016-03-22    0.021373
2016-03-23    0.018532
2016-03-24    0.019767
2016-03-25    0.019211
2016-03-26    0.016802
2016-03-27    0.015649
2016-03-28    0.020859
2016-03-29    0.022341
2016-03-30    0.024771
2016-03-31    0.023783
2016-04-01    0.022794
2016-04-02    0.024915
2016-04-03    0.025203
2016-04-04    0.024483
2016-04-05    0.124761
2016-04-06    0.221806
2016-04-07    0.131947
Name: last_seen, dtype: float64

Observations:

* The 'last_seen' dates indicate the last time the ad was seen by the crawler.
* The distribution is right-skewed, with a significant proportion of ads being seen in the last few days of the dataset period (April 5, April 6, and April 7, 2016).
* This suggests a high level of activity in terms of ad views towards the end of the dataset period, possibly indicating a surge in user activity or the removal of listings from the platform.

In [27]:
# Generate descriptive statistics for the 'registration_year' column
autos['registration_year'].describe()

count    48565.000000
mean      2004.755421
std         88.643887
min       1000.000000
25%       1999.000000
50%       2004.000000
75%       2008.000000
max       9999.000000
Name: registration_year, dtype: float64

These observations suggest that while most vehicles were registered in the early to mid-2000s, there are anomalies in the data, including outliers and potential data entry errors, which may require further investigation and cleaning.

# Dealing with Incorrect Registration Year Data

One thing that stands out from the exploration we did in the last screen is that the registration_year column contains some odd values:

* The minimum value is 1000, before cars were invented
* The maximum value is 9999, many years into the future

Because a car can't be first registered after the listing was seen, any vehicle with a registration year above 2016 is definitely inaccurate. Determining the earliest valid year is more difficult. Realistically, it could be somewhere in the first few decades of the 1900s.

Let's count the number of listings with cars that fall outside the 1900 - 2016 interval and see if it's safe to remove those rows entirely, or if we need more custom logic.

In [28]:
# Counting the number of listings with registration years outside the 1900 - 2016 interval
out_of_range_count=((autos['registration_year']<1900) | (autos['registration_year']>2016)).sum()
out_of_range_count

1884

Observations:
* Highest Acceptable Value: The highest acceptable value for the registration_year column should be set around the current year (2016) or a few years beyond it, ensuring that it does not extend into the future.

* Lowest Acceptable Value: The lowest acceptable value should be set in the early 1900s, considering the advent of automobile manufacturing and standardized registration processes, ensuring that it does not fall before this historical context.

These boundaries help filter out erroneous or outlier registration years, ensuring the data remains realistic and valid within the context of automobile registration history.







In [29]:
# Filtering the dataset to include only registration years within the acceptable range
autos=autos[autos['registration_year'].between(1900,2016)]

In [30]:
autos['registration_year'].describe()

count    46681.000000
mean      2002.910756
std          7.185103
min       1910.000000
25%       1999.000000
50%       2003.000000
75%       2008.000000
max       2016.000000
Name: registration_year, dtype: float64

In [31]:
# Calculating the distribution of the remaining registration years
registration_year_distribution=autos['registration_year'].value_counts(normalize=True,dropna=False).sort_index()
registration_year_distribution

1910    0.000107
1927    0.000021
1929    0.000021
1931    0.000021
1934    0.000043
          ...   
2012    0.028063
2013    0.017202
2014    0.014203
2015    0.008397
2016    0.026135
Name: registration_year, Length: 78, dtype: float64

Observations:
* The distribution is concentrated in the early 20th century and recent years.
* There's a gradual increase in frequency from the early 1900s to the mid-20th century.
* Higher frequencies are observed for registration years from the late 20th century to the early 21st century.
* The peak occurs around the years 2000 to 2010.
* There's a decline in frequency for registration years beyond 2010.






# Exploring Price by Brand

To explore the variations in price across different car brands, we can use aggregation

In [32]:
# Calculate the frequency of each brand
brands= autos['brand'].value_counts(normalize=True,dropna=False)
brands

volkswagen        0.211264
bmw               0.110045
opel              0.107581
mercedes_benz     0.096463
audi              0.086566
ford              0.069900
renault           0.047150
peugeot           0.029841
fiat              0.025642
seat              0.018273
skoda             0.016409
nissan            0.015274
mazda             0.015188
smart             0.014160
citroen           0.014010
toyota            0.012703
hyundai           0.010025
sonstige_autos    0.009811
volvo             0.009147
mini              0.008762
mitsubishi        0.008226
honda             0.007840
kia               0.007069
alfa_romeo        0.006641
porsche           0.006127
suzuki            0.005934
chevrolet         0.005698
chrysler          0.003513
dacia             0.002635
daihatsu          0.002506
jeep              0.002271
subaru            0.002142
land_rover        0.002099
saab              0.001649
jaguar            0.001564
daewoo            0.001500
trabant           0.001392
r

In [33]:
# Select brands with a percentage greater than 5%
selected_brands = brands[brands>0.05].index
selected_brands

Index(['volkswagen', 'bmw', 'opel', 'mercedes_benz', 'audi', 'ford'], dtype='object')

After exploring the brand data in the dataset, it was observed that there are numerous unique brands represented. To focus our analysis, we decided to aggregate on brands that have a significant presence in the dataset, selecting those with a percentage of occurrences greater than 5%. This approach allows us to concentrate on brands that are relatively more common in the dataset and likely have a more significant impact on the overall analysis.

In [34]:
# creating and empty dictionary to hold aggregate data
brand_mean_prices={}

In [35]:
# loop over selected brands and calculate the mean price
for brands in selected_brands:
    brand_only=autos[autos['brand']==brands]
    mean_price=brand_only['price'].mean()
    brand_mean_prices[brands]=int(mean_price)

In [36]:
# Print the dictionary of aggregate data
brand_mean_prices

{'volkswagen': 5402,
 'bmw': 8332,
 'opel': 2975,
 'mercedes_benz': 8628,
 'audi': 9336,
 'ford': 3749}

Audi tops the list with the highest average price, followed closely by BMW and Mercedes-Benz. Volkswagen comes next with a decent average, while Opel and Ford are on the lower end. This suggests that luxury brands like Audi and BMW tend to have higher prices compared to more mainstream brands like Volkswagen and Ford.

# Storing Aggregate Data in a DataFrame


To create a DataFrame from the aggregated data, we first use the pandas Series constructor with our dictionary of brand mean prices. This constructor converts our dictionary into a pandas Series, where the keys become the index labels.

In [37]:
bmp_series=pd.Series(brand_mean_prices)
bmp_series

volkswagen       5402
bmw              8332
opel             2975
mercedes_benz    8628
audi             9336
ford             3749
dtype: int64

Then, we use the pandas DataFrame constructor to create a DataFrame from this Series, specifying the column name with the columns parameter. This way, we can create a DataFrame with one column representing the mean prices of each brand.

In [38]:
bmp_df=pd.DataFrame(bmp_series,columns=['mean_price'])
bmp_df

mean_price
volkswagen           5402
bmw                  8332
opel                 2975
mercedes_benz        8628
audi                 9336
ford                 3749

In [39]:
# Calculating mean mileage

# creating and empty dictionary to hold aggregate data
brand_mean_mileage={}

# loop over selected brands and calculate the mean price
for brands in selected_brands:
    brand_only=autos[autos['brand']==brands]
    mean_mileage=brand_only['odometer_km'].mean()
    brand_mean_mileage[brands]=int(mean_mileage)

# printing the dictionary of aggregate data
brand_mean_mileage

{'volkswagen': 128707,
 'bmw': 132572,
 'opel': 129310,
 'mercedes_benz': 130788,
 'audi': 129157,
 'ford': 124266}

In [40]:
# Convert dictionaries to series objects
mean_price=pd.Series(brand_mean_prices).sort_values(ascending=False)
mean_mileage=pd.Series(brand_mean_mileage).sort_values(ascending=False)

In [41]:
# Create a DataFrame from the series object
top_brand_info=pd.DataFrame(mean_price,columns=['mean_price'])
top_brand_info

mean_price
audi                 9336
mercedes_benz        8628
bmw                  8332
volkswagen           5402
ford                 3749
opel                 2975

In [42]:
# Assign the new series as a new column in the existing DataFrame
top_brand_info['mean_mileage']=mean_mileage
top_brand_info

mean_price  mean_mileage
audi                 9336        129157
mercedes_benz        8628        130788
bmw                  8332        132572
volkswagen           5402        128707
ford                 3749        124266
opel                 2975        129310

The DataFrame shows the average prices and mileage for the most popular car brands. Expensive brands like Audi, BMW, and Mercedes have higher prices, but they also tend to have higher mileage. On the other hand, cheaper brands like Ford and Opel have lower prices and lower mileage. Volkswagen is in the middle, both in terms of price and mileage. This means that while luxury brands cost more, mileage doesn't always directly affect their prices.

Let's continue to clean up the data. To do this we'll indentify the categorical data that uses german words and translate and map the values to their english counterparts.

In [43]:
autos.head()

date_crawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_T�V_neu_ist_sehr_gepfleg...   

    price   abtest vehicle_type  registration_year    gearbox  power_ps  \
0  5000.0  control          bus               2004    manuell       158   
1  8500.0  control    limousine               1997  automatik       286   
2  8990.0     test    limousine               2009    manuell       102   
3  4350.0  control   kleinwagen               2007  automatik        71   
4  1350.0     test        kombi               2003    manuell         0   

    model  odometer_km  registration_month fuel_type       brand  \
0  andere     150000.0                   3       lpg     peugeot   
1     7er     150000.0                   6    benzin         bmw   
2    golf      70000.0                   7    benzin  volkswagen   
3  fortwo      70000.0                   6    benzin       smart   
4   focus     150000.0                   7    benzin        ford   

  unrepaired_damage           ad_created  postal_code            last_seen  
0              nein  2016-03-26 00:00:00        79588  2016-04-06 06:45:54  
1              nein  2016-04-04 00:00:00        71034  2016-04-06 14:45:08  
2              nein  2016-03-26 00:00:00        35394  2016-04-06 20:15:37  
3              nein  2016-03-12 00:00:00        33729  2016-03-15 03:16:28  
4              nein  2016-04-01 00:00:00        39218  2016-04-01 14:38:50

It looks like data in German is presented in the following columns:

* vehicle_type
* gearbox
* fuel_type
* unrepaired_damage

Let's explore each column.



In [44]:
# Identifying the German words in 'gearbox' columns
autos['gearbox'].unique()

array(['manuell', 'automatik', nan], dtype=object)

In [45]:
# Identifying the German words in 'unrepaired_damage' columns
autos['unrepaired_damage'].unique()

array(['nein', nan, 'ja'], dtype=object)

In [46]:
# Identifying the German words in 'fuel_type' columns
autos['fuel_type'].unique()

array(['lpg', 'benzin', 'diesel', nan, 'cng', 'hybrid', 'elektro',
       'andere'], dtype=object)

In [47]:
# Identifying the German words in 'vehicle_type' columns
autos['vehicle_type'].unique()

array(['bus', 'limousine', 'kleinwagen', 'kombi', nan, 'coupe', 'suv',
       'cabrio', 'andere'], dtype=object)

In order to translate the German words in the columns mentioned above, I will create a dictionary, with German words as keys and their English translations as values

In [48]:
# Mapping dictionary for translations
mapping_dic={
    'manuell': 'manual',
    'automatik': 'automatic',
    'nein': 'no',
    'ja': 'yes',
    'benzin': 'petrol',
    'diesel': 'diesel',
    'lpg': 'lpg',
    'cng': 'cng',
    'hybrid': 'hybrid',
    'elektro': 'electric',
    'andere': 'other',
    'limousine': 'sedan',
    'kleinwagen': 'small car',
    'kombi': 'station wagon',
    'bus': 'bus',
    'cabrio': 'convertible',
    'coupe': 'coupe',
    'suv': 'suv',
    'andere': 'other'
}

In [49]:
# Specify the columns to be translated
cols_translated = ['gearbox', 'fuel_type', 'unrepaired_damage','vehicle_type']

In [50]:
# Iterate over the columns and map their values using the provided mapping dictionary
for col in cols_translated:
    autos[col] = autos[col].map(mapping_dic)

Checking the result:

In [51]:
print(autos['vehicle_type'].unique())
print(autos['gearbox'].unique())
print(autos['fuel_type'].unique())
print(autos['unrepaired_damage'].unique())

['bus' 'sedan' 'small car' 'station wagon' nan 'coupe' 'suv' 'convertible'
 'other']
['manual' 'automatic' nan]
['lpg' 'petrol' 'diesel' nan 'cng' 'hybrid' 'electric' 'other']
['no' nan 'yes']


# Convert the dates to be uniform numeric data, so "2016-03-21" becomes the integer 20160321.

Let's identify columns in the dataset that contain dates.

In [52]:
autos.head()

date_crawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_T�V_neu_ist_sehr_gepfleg...   

    price   abtest   vehicle_type  registration_year    gearbox  power_ps  \
0  5000.0  control            bus               2004     manual       158   
1  8500.0  control          sedan               1997  automatic       286   
2  8990.0     test          sedan               2009     manual       102   
3  4350.0  control      small car               2007  automatic        71   
4  1350.0     test  station wagon               2003     manual         0   

    model  odometer_km  registration_month fuel_type       brand  \
0  andere     150000.0                   3       lpg     peugeot   
1     7er     150000.0                   6    petrol         bmw   
2    golf      70000.0                   7    petrol  volkswagen   
3  fortwo      70000.0                   6    petrol       smart   
4   focus     150000.0                   7    petrol        ford   

  unrepaired_damage           ad_created  postal_code            last_seen  
0                no  2016-03-26 00:00:00        79588  2016-04-06 06:45:54  
1                no  2016-04-04 00:00:00        71034  2016-04-06 14:45:08  
2                no  2016-03-26 00:00:00        35394  2016-04-06 20:15:37  
3                no  2016-03-12 00:00:00        33729  2016-03-15 03:16:28  
4                no  2016-04-01 00:00:00        39218  2016-04-01 14:38:50

There are three columns with dates in them: 
* date_crawled
* ad_created
* last_seen

Before converting the dates to numeric, I will have to extract the dates from each column. I see that data in all three columns has identical format and it will make my work easier.

For each column I will:
* extract first 10 symbols
* clean the data from "-" sign
* convert the remained data to numeric

In [55]:
date_cols=['date_crawled','ad_created','last_seen']
for each in date_cols:
    autos[each]=(autos[each].str[:10].str.replace('-','').astype(int))

In [56]:
autos.head()

date_crawled                                               name   price  \
0      20160326                   Peugeot_807_160_NAVTECH_ON_BOARD  5000.0   
1      20160404         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik  8500.0   
2      20160326                         Volkswagen_Golf_1.6_United  8990.0   
3      20160312  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...  4350.0   
4      20160401  Ford_Focus_1_6_Benzin_T�V_neu_ist_sehr_gepfleg...  1350.0   

    abtest   vehicle_type  registration_year    gearbox  power_ps   model  \
0  control            bus               2004     manual       158  andere   
1  control          sedan               1997  automatic       286     7er   
2     test          sedan               2009     manual       102    golf   
3  control      small car               2007  automatic        71  fortwo   
4     test  station wagon               2003     manual         0   focus   

   odometer_km  registration_month fuel_type       brand unrepaired_damage  \
0     150000.0                   3       lpg     peugeot                no   
1     150000.0                   6    petrol         bmw                no   
2      70000.0                   7    petrol  volkswagen                no   
3      70000.0                   6    petrol       smart                no   
4     150000.0                   7    petrol        ford                no   

   ad_created  postal_code  last_seen  
0    20160326        79588   20160406  
1    20160404        71034   20160406  
2    20160326        35394   20160406  
3    20160312        33729   20160315  
4    20160401        39218   20160401

# Exploring Popular Brand/Model Combinations


In [59]:
# Grouping by brand and then get the value counts for the models in each brand
autos.groupby('brand')['model'].value_counts().sort_values(ascending=False).head(10)

brand          model   
volkswagen     golf        3707
bmw            3er         2615
volkswagen     polo        1609
opel           corsa       1592
volkswagen     passat      1349
opel           astra       1348
audi           a4          1231
mercedes_benz  c_klasse    1136
bmw            5er         1132
mercedes_benz  e_klasse     958
Name: model, dtype: int64

# Split the odometer_km into groups, and use aggregation to see if average prices follows any patterns based on the mileage.

To split the odometer_km into groups and analyze average prices based on mileage, we'll follow these steps:

* Define mileage groups based on the odometer_km values.
* Group the dataset by these mileage groups.
* Calculate the average price for each mileage group.

In [65]:
# Define mileage groups
mileage_groups = pd.cut(autos['odometer_km'], bins=[0, 50000, 100000, 150000], 
                        labels=['<50k', '50k-100k', '100k-150k'])
mileage_groups

0        100k-150k
1        100k-150k
2         50k-100k
3         50k-100k
4        100k-150k
           ...    
49995     50k-100k
49996    100k-150k
49997         <50k
49998         <50k
49999    100k-150k
Name: odometer_km, Length: 46681, dtype: category
Categories (3, object): ['<50k' < '50k-100k' < '100k-150k']

In [66]:
# Group by mileage groups and calculate average price
average_price_by_mileage = autos.groupby(mileage_groups)['price'].mean()
average_price_by_mileage

odometer_km
<50k         14890.828856
50k-100k      9595.419350
100k-150k     4107.938641
Name: price, dtype: float64

# How much cheaper are cars with damage than their non-damaged counterparts?

In [69]:
autos.groupby(['unrepaired_damage']).price.mean().sort_values(ascending=False)

unrepaired_damage
no     7164.033103
yes    2241.146035
Name: price, dtype: float64

We can see that cars with unrepaired damage are significantly cheaper on average than their non-damaged counterparts.

 # Conclusion
* The project involved cleaning and analyzing a dataset of used cars from eBay Kleinanzeigen.
* Tasks included renaming columns for clarity, handling missing values, and converting data types.
* German words in categorical columns were translated to English for better understanding.
* Analysis revealed insights into the distribution of car prices, mileage, and popular brands.
* A notable finding was that cars with unrepaired damage are significantly cheaper on average than their non-damaged counterparts.
* The project demonstrated the importance of data cleaning and exploratory analysis in extracting meaningful insights from datasets.